### PR TITLE
WIP: Have plotting functions operate on axis

### PR DESCRIPTION
**Pull request by mgeplf**
* (neuron, neron3d, tree, tree3d, soma, soma3d, dendrite) in
  neurom/view/view.py all operate on matplotlib.Axes, so the user
  has more control over what is happening, and can compose plots
  more simply.

* They generally set the dataLim, so axes views can be inferred
  properly

### CODE REVIEW

**Comment by mgeplf** [comment]
Currently most of the tests are turned off.

The old neuron.viewer.draw works the same:
Old:
old
New:
new

I think the 3d view looks better (but I'm still having issues w/ a locked aspect ratio):
Old:
original_3d

New:
new_3d

The new code means it's easy to have good control over everything:
```
fig = plt.figure()
planes = ('xy', 'zy')
for i, plane in enumerate(planes):
    ax = fig.add_subplot(1, len(planes), i + 1)
    v.neuron(ax, m, plane=plane)
    common.plot_style(fig=fig, ax=ax, white_space=30)
    ax.set_title(plane)
    fig.set_tight_layout(True)
```
Results in:
multi

Which was possible before, but I think this is more clear, and offers more control over the axis.  Again, this is the develop API, so I think we should assume they are knowledgeable about matplotlib.

@lidakanari @eleftherioszisis thoughts?

**Comment by lidakanari** [comment]
Looks good! Last part of your code is equivalent to:
```
common.plot_style(fig=fig, ax=ax, white_space=30, title=plane, tight=True)
```
I think the functionality is the same but it is probably easier to understand it :) -if you are of course knowledgeable about matplotlib-.

Is it possible to add back the control over the type of neurites that are plotted? For instance to choose to plot only axons.

File: examples/plot_somas.py [hunk 1]
```diff
@@ -46,12 +47,11 @@ def random_color():
...
     '''Plot set of somas on same figure as spheres, each with different color'''
-    fig, ax = common.get_figure(new_fig=True, subplot=111,
-                                params={'projection': '3d', 'aspect': 'equal'})
+    _, ax = common.get_figure(new_fig=True, subplot=111,
+                              params={'projection': '3d', 'aspect': 'equal'})
     for s in somas:
         common.plot_sphere(ax, s.center, s.radius, color=random_color(), alpha=1)
-    fig.show()
-    plt.close(fig)
+    plt.show()
```
Comment (mgeplf) [inline examples/plot_somas.py]: Changed so figure shows up until it's closed..

File: neurom/core/_neuron.py [hunk 2]
```diff
@@ -212,6 +212,11 @@ def __deepcopy__(self, memo):
         '''Deep copy of neurite object'''
         return Neurite(deepcopy(self.root_node, memo))
 
+    def __nonzero__(self):
```
Comment (mgeplf) [inline neurom/core/_neuron.py]: Make it simple to test if a tree/neurite/neuron is empty

File: neurom/tests/test_viewer.py [hunk 3]
```diff
@@ -27,6 +27,12 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import os
+import shutil
+import tempfile
+
+import matplotlib
```
Comment (mgeplf) [inline neurom/tests/test_viewer.py]: Saves us from Agg being set by view.common for everyone.

File: neurom/view/common.py [hunk 4]
```diff
@@ -26,20 +26,9 @@
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-"""
-Module containing the common functionality
-to be used by view-plot modules.
-"""
-from neurom import NeuriteType
+"""Functionality for styling plots"""
```
Comment (mgeplf) [inline neurom/view/common.py]: Most of the changes in this file have to do w/:
* remove kwargs when possible, make all args explicit: better for completion, better for error handling, easier to write docstrings for

File: neurom/view/view.py [hunk 5]
```diff
@@ -25,13 +25,12 @@
 # ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-'''
-Python module of NeuroM to visualize morphologies
-'''
+'''visualize morphologies'''
```
Comment (mgeplf) [inline neurom/view/view.py]: Changes have to do with:
* removing kwargs
* removing automatic figure creation: this is a development API, and it's better to give complete control to the person using matplotlib
* remove automatic styling, again, a developer can call this if necessary
* rename functions neuron -> plot_neuron

File: neurom/viewer.py [hunk 6]
```diff
@@ -112,6 +121,16 @@ def draw(obj, mode='2d', **kwargs):
 
     viewer = '%s_%s' % (tag, mode)
     try:
-        return _VIEWERS[viewer](obj, **kwargs)
+        plotter = _VIEWERS[viewer]
+        output_path = kwargs.pop('output_path', False)
```
Comment (lidakanari) [inline neurom/viewer.py]: None should have the same effect, right?

File: neurom/view/common.py [hunk 7]
```diff
@@ -50,542 +39,311 @@
...
     if zticks is not None:
         ax.set_zticks(zticks)
-        ax.zaxis.set_tick_params(labelsize=tick_fontsize, **zticks_arg)
+        zticks_args = dict_if_none(zticks_args)
+        ax.zaxis.set_tick_params(labelsize=tick_fontsize, **zticks_args)
 
 
-def plot_limits(ax, **kwargs):
+def plot_limits(ax, white_space):
     """Sets the limit options of a matplotlib plot.
```
Comment (lidakanari) [inline neurom/view/common.py]: Shouldn't we also control if someone wants to pass the limits directly? Sorry if this is done somewhere else but this review is too long and my memory limited :P

File: neurom/view/tests/utils.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+from neurom.view.common import plt
+
+
+@contextmanager
+def get_fig_2d():
+    fig = plt.figure()
+    ax = fig.add_subplot(111)
+    ax.plot([0, 0], [1, 2], label='test')
+    yield fig, ax
+    plt.close(fig)
```
Comment (mgeplf) [inline neurom/view/tests/utils.py]: Good point, I will fix that.

File: neurom/viewer.py (hunk 6 above)
Comment (mgeplf) [inline neurom/viewer.py]: fixed

File: neurom/viewer.py [hunk 9]
```diff
@@ -112,6 +121,16 @@ def draw(obj, mode='2d', **kwargs):
...
+        plotter(ax, obj, **kwargs)
+
+        if mode != 'dendrogram':
+            common.plot_style(fig=fig, ax=ax, **kwargs)
+
+        if output_path:
+            common.save_plot(fig=fig, output_path=output_path, **kwargs)
+
+        return fig, ax
     except KeyError:
```
Comment (mgeplf) [inline neurom/viewer.py]: agreed

File: neurom/view/common.py (hunk 7 above)
Comment (mgeplf) [inline neurom/view/common.py]: Yeah, it's huge, and I apologize for that.

I should rename this function to more clearly state what it does, which I will do.  If we want another function to directly set the limits, I can add that, it just probably wouldn't be called by any of the internal functions, so I don't know if it's something people would use

Comment (lidakanari) [inline neurom/view/common.py]: OK, it is certainly trivial to call the limits, so I don't mind not including this.

**Pull request merged**